Q1: 

A1.Download and install Gutenberg corpus tool to your Jupyter Notebook. Project Gutenberg contains some 25,000 free electronic books, hosted at http://www.gutenberg.org/. We can install the NLTK package, then use the Gutenberg corpus in it.  Can be installed by running the following in computer terminal: 

In [1]:
pip install nltk

Note: you may need to restart the kernel to use updated packages.


B.Download Gutenberg corpus tool in NLTK package by e.g.:

In [80]:
import nltk.corpus
nltk.download('gutenberg')

[nltk_data] Downloading package gutenberg to
[nltk_data]     /Users/animeshsrivastava/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!


True

In [81]:
import pandas as pd
import numpy as np
from nltk import FreqDist
from nltk.corpus import wordnet as wn
import nltk
from nltk.corpus import inaugural

In [82]:
from nltk.probability import ConditionalFreqDist

C.Use the texts in the corpus.

In [83]:
nltk.corpus.gutenberg.fileids()

['austen-emma.txt',
 'austen-persuasion.txt',
 'austen-sense.txt',
 'bible-kjv.txt',
 'blake-poems.txt',
 'bryant-stories.txt',
 'burgess-busterbrown.txt',
 'carroll-alice.txt',
 'chesterton-ball.txt',
 'chesterton-brown.txt',
 'chesterton-thursday.txt',
 'edgeworth-parents.txt',
 'melville-moby_dick.txt',
 'milton-paradise.txt',
 'shakespeare-caesar.txt',
 'shakespeare-hamlet.txt',
 'shakespeare-macbeth.txt',
 'whitman-leaves.txt']

D.Create a table displaying relative frequencies with which “modals” (can, could, may, might, will, would and should) are used in all texts provided in the corpus.

In [69]:
words = []
for fileid in nltk.corpus.gutenberg.fileids():
    for w in nltk.corpus.gutenberg.words(fileid):
        words.append(w)
len(words)

2621613

In [70]:
modals = ['can','could','may','might','will','would','should']
filtered_words = [word for word in words if word in modals]
FreqDist(filtered_words).tabulate()

  will  would  could should    may    can  might 
  7130   3932   3528   2496   2435   2163   1938 


E.For two modals with the largest span of relative frequencies (most used minus least used), select a text which uses it the most and the text that uses it the least. Compare usage in both texts by examining the relative frequencies of those modals in two texts. Try to suggest an explanation why  those words used differently in the two texts.

In [84]:
rfd= nltk.ConditionalFreqDist(
    (files,word)
    for files in gutenberg.fileids()
    for word in gutenberg.words(files))
modals = ['might','will']
rfd.tabulate(conditions=gutenberg.fileids(), samples=modals)

                        might  will 
        austen-emma.txt   322   559 
  austen-persuasion.txt   166   162 
       austen-sense.txt   215   354 
          bible-kjv.txt   475  3807 
        blake-poems.txt     2     3 
     bryant-stories.txt    23   144 
burgess-busterbrown.txt    17    19 
      carroll-alice.txt    28    24 
    chesterton-ball.txt    69   198 
   chesterton-brown.txt    71   111 
chesterton-thursday.txt    71   109 
  edgeworth-parents.txt   127   517 
 melville-moby_dick.txt   183   379 
    milton-paradise.txt    98   161 
 shakespeare-caesar.txt    12   129 
 shakespeare-hamlet.txt    28   131 
shakespeare-macbeth.txt     5    62 
     whitman-leaves.txt    26   261 


According to the above result the two modals with the largest relative frequencies are - "will" and "might". In both the modals, bible-kjv.txt uses the most with relative frequencies of 3807 and 475. In both the modals, blake-poems.txt uses the least with relative frequencies of 3 and 2. It is very clear that bible wil have the most usage of these modals as they connates occurance of any event that is going to happen. Bible is also written mostly in future tense. These modals are least used in blake poems because they mostly signify romanticism and the writing style contains less modals due to rhyming nature.

Q2: 

A2.In the Inaugural corpus, see below:

In [85]:
import nltk
nltk.download('inaugural')

[nltk_data] Downloading package inaugural to
[nltk_data]     /Users/animeshsrivastava/nltk_data...
[nltk_data]   Package inaugural is already up-to-date!


True

B.Chose Kennedy’s speech, using e.g. this code:

In [86]:
np_words = nltk.corpus.inaugural.words('1961-Kennedy.txt')

C.Identify the 10 most frequently used long words (words longer than 7 characters).

In [87]:
filt_char = ([w.lower() for w in np_words if len(w) > 7])

# Frequency distribution
f_dist = FreqDist(filt_char)
freq = f_dist.most_common(10)

# Creaty empty arrays
np_1 = []
np_2 = []

# Seperate words
for i in range(10):
    np_1.append(freq[i][0])

for i in range(10):
    np_2.append(freq[i][1])

# Create dataframe
df_freq = pd.DataFrame({'Top 10':np_1, 'n': np_2})

In [88]:
df_freq

Top 10  n
0    citizens  5
1   president  4
2   americans  4
3  generation  3
4   forebears  2
5  revolution  2
6   committed  2
7    powerful  2
8  supporting  2
9  themselves  2

D.Which one of those 10 words has the largest number of synonyms? Use WordNet as a helper:

In [89]:
np_syn = []

# Loop trough distribution
for i in f_dist.most_common(10):
    # Print word
    print( 'Word: ', i[0])
    count=0
    
    # Print synonym
    print ('Synonymns: ')
    for j in wn.synsets(i[0]):
        print ('  ',j.lemma_names())
        count+=len(j.lemma_names())
    np_syn.append([i[0], count])  

Word:  citizens
Synonymns: 
   ['citizen']
Word:  president
Synonymns: 
   ['president']
   ['President_of_the_United_States', 'United_States_President', 'President', 'Chief_Executive']
   ['president']
   ['president', 'chairman', 'chairwoman', 'chair', 'chairperson']
   ['president', 'prexy']
   ['President_of_the_United_States', 'President', 'Chief_Executive']
Word:  americans
Synonymns: 
   ['American']
   ['American_English', 'American_language', 'American']
   ['American']
Word:  generation
Synonymns: 
   ['coevals', 'contemporaries', 'generation']
   ['generation']
   ['generation']
   ['generation']
   ['genesis', 'generation']
   ['generation']
   ['generation', 'multiplication', 'propagation']
Word:  forebears
Synonymns: 
   ['forebear', 'forbear']
Word:  revolution
Synonymns: 
   ['revolution']
   ['revolution']
   ['rotation', 'revolution', 'gyration']
Word:  committed
Synonymns: 
   ['perpetrate', 'commit', 'pull']
   ['give', 'dedicate', 'consecrate', 'commit', 'devote']


In [96]:
# Creaty empty arrays
np_1 = []
np_2 = []

# Get synonyms
for i in range(10):
    np_1.append(np_syn[i][0])

for j in range(10):
    np_2.append(np_syn[j][1])

# Create dataframe
df_syn = pd.DataFrame({'Top 10 synonyms':np_1, 'n':np_2})

E.List all synonyms for the 10 most frequently used words. Which one of those 10 words has the largest number of hyponyms? 

In [97]:
df_syn.sort_values(by="n", ascending=False)

Top 10 synonyms   n
8      supporting  52
6       committed  27
1       president  16
7        powerful  16
3      generation  12
2       americans   5
5      revolution   5
4       forebears   2
0        citizens   1
9      themselves   0

In [93]:
np_hyp = []

# Loop trough distribution
for i in f_dist.most_common(10):
    # Print word
    print ('Word: ', i[0])
    count=0
    
    # Print Hyponyms
    print ('Hyponyms: ')
    for j in wn.synsets(i[0]):
        print (j)
        print (j.hyponyms())
        count+= len(j.hyponyms())
    np_hyp.append([i[0], count])

Word:  citizens
Hyponyms: 
Synset('citizen.n.01')
[Synset('active_citizen.n.01'), Synset('civilian.n.01'), Synset('freeman.n.01'), Synset('private_citizen.n.01'), Synset('repatriate.n.01'), Synset('thane.n.02'), Synset('voter.n.01')]
Word:  president
Hyponyms: 
Synset('president.n.01')
[]
Synset('president_of_the_united_states.n.01')
[]
Synset('president.n.03')
[Synset('ex-president.n.01')]
Synset('president.n.04')
[Synset('kalon_tripa.n.01'), Synset('vice_chairman.n.01')]
Synset('president.n.05')
[]
Synset('president_of_the_united_states.n.02')
[]
Word:  americans
Hyponyms: 
Synset('american.n.01')
[Synset('african-american.n.01'), Synset('alabaman.n.01'), Synset('alaskan.n.01'), Synset('anglo-american.n.01'), Synset('appalachian.n.01'), Synset('arizonan.n.01'), Synset('arkansan.n.01'), Synset('asian_american.n.01'), Synset('bay_stater.n.01'), Synset('bostonian.n.01'), Synset('californian.n.01'), Synset('carolinian.n.01'), Synset('coloradan.n.01'), Synset('connecticuter.n.01'), Synset

In [98]:
np_1 = []
np_2 = []

# Get synonyms
for i in range(10):
    np_1.append(np_hyp[i][0])

for j in range(10):
    np_2.append(np_hyp[j][1])

# Create dataframe
df_hyp = pd.DataFrame({'Top 10 Hyponyms':np_1, 'n':np_2})

F.List all hyponyms of the 10 most frequently used words.

In [99]:
df_hyp.sort_values(by="n", ascending=False)

Top 10 Hyponyms   n
2       americans  75
8      supporting  42
6       committed  16
5      revolution   8
0        citizens   7
3      generation   6
1       president   3
4       forebears   2
7        powerful   0
9      themselves   0

G.Reflect on the results.

The word with highest number of synonyms is "Supporting". The meaning of support is to promote interest or cause of any activity. There are many words with that type of meaning. The word "Themselves" and "citizens" have the least synonyms as they relate to a particular type with not much of diversity. A word whose meaning is included in the meaning of another more general word is known as Hyponyms. The word " Americans have the most number of hyponyms as the americans have many types and its meaning is related to many general word. The word "Themselves" and "Powerful" has no hyponyms as they are in their general form.